### ISSUE TITLE
No -o option for xargs in Termux

### ISSUE THREAD

**Issue by jarun** [jarun/nnn#387]
We have a common `xargs -o` in the code in `cpmv_rename()`.

We need to work around this has Termux does not have a `-o` option for `xargs`.

@KlzXS please take a look.

**Comment by KlzXS** [comment]
Which approach are you for, completely eliminating it or finding a way to detect if it's present?

**Comment by jarun** [comment]
If we can eliminate it, that would be the best.

**Comment by KlzXS** [comment]
Should I also remove the flag from `opstr` and `rmmulstr`. What about the `-a` flag to get rid of the `#ifdef __linux__`?

**Comment by jarun** [comment]
If we can do that, it's the best!

The non-Linux cases are *BSD and macOS, both of which have the `-o` option. But if that works without `-o` we can be uniform and get rid of the `#ifdef __linux__`.

For what I have seen so far, the BusyBox `xargs` is the minimal (to the point of being crippled(?)) followed by Termux (no `-o`).

**Comment by jarun** [comment]
`-a` is not available on *BSD and macOS:
- https://man.openbsd.org/xargs.1
- https://ss64.com/osx/xargs.html

I think what we really want to do is make the Linux part work on Termux and BusyBox.

For the common cases, we have to get rid of `-o`. But where we need to confirm per file, we should (like `rm` multiple).

**Comment by KlzXS** [comment]
Well I can get rid of `-a` easily by just redirecting a file as input to `xargs`. As for `-o` I can do what it does manually by making `xargs` run not the command itself but start a shell which will run a command and redirect its `stdin` to `/dev/tty`. I'll try to find a way to do it without spawning a new shell.

This should keep the functionality just like it was on systems which support `-a` and `-o` but also make it work for system which don't.

**Comment by jarun** [comment]
> Well I can get rid of -a easily by just redirecting a file as input to xargs.

That's what I tried in commit 407b690404e60f3c66a8b7bb2b580b8d2e6e61f3. Both BusyBox and Termux will work.

> I'll try to find a way to do it without spawning a new shell.

If we can detect if `xargs` doesn't have `-o` we can try shell spawning (or any other better optimize approach). If it's easy the capability detection will reduce our changes.

Can we detect by running `echo "x" | xargs -o` output? We should do this only once and set a flag in settings. Just check it out. Or if you know of a simpler mechanism even better.

The other thing BusyBox complained about is `-I` which is coming from `REPLACE_STR`. I am not sure what we need to do for this.

**Comment by KlzXS** [comment]
> That's what I tried in commit 407b690. Both BusyBox and Termux will work

This but without cat abuse. We can just redirect a file `<file xargs ...`

> Can we detect by running echo "x" | xargs -o output? We should do this only once and set a flag in settings. Just check it out. Or if you know of a simpler mechanism even better.

That would work. If any error is encountered a non-zero value is returned.

>The other thing BusyBox complained about is -I which is coming from REPLACE_STR. I am not sure what we need to do for this.

`-I` is only used in `opstr()`. If we are spawning a shell for this we can remove it. The shell can expand arguments passed to it with `$@` and `$0` (usually this would be the command being run, but with `sh -c` you can set it to whatever you like).

**Comment by jarun** [comment]
> This but without cat abuse.

Nice!

> That would work.

Awesome!

> If we are spawning a shell for this we can remove it.

If we can detect and spawn a shell only if `-I` is not available that would be great.

**Comment by KlzXS** [comment]
So to sum up. Remove all and any `-a`. Detect (non)existance of `-I` and `-o`, use them if available, otherwise spawn a if a better way is not found.

Where should the flags be checked and where should I store the result? The settings struct?

**Comment by jarun** [comment]
That sums it up right!

The flags should be checked once at start and set in a flag in settings struct.

**Comment by jarun** [comment]
Adding @BKPepe in the discussion. We are working on having support for BusyBox. Please have patience.

**Comment by jarun** [comment]
Adding @BKPepe in the discussion. We are working on having support for BusyBox. Please have patience.

**Comment by KlzXS** [comment]
So I've decided we are going with spawning the shell approach everywhere and not using `-I` and `-o`, especially for `opstr()`. It's not even funny.

Tested with 100'000 4KiB files containing output of `/dev/urandom`.
Ran each option a few times, similar results to ones displayed below.
```
time <list xargs -0 -o -I{} cp {} nnn/
real    1m15.940s
user    0m58.626s
sys     0m20.784s

time <list xargs -0 -I{} cp {} nnn/
real    1m17.515s
user    0m59.171s
sys     0m21.253s

time <list xargs -0 -o sh -c 'cp $0 $@ nnn/'
real    0m2.628s
user    0m0.447s
sys     0m2.134s

time <list xargs -0 sh -c 'cp $0 $@ nnn/ </dev/tty'
real    0m2.662s
user    0m0.444s
sys     0m2.192s

```
`mv` is also shows similar results which is ridiculous. 
I would like for someone to replicate the findings.

**Comment by jarun** [comment]
@0xACE can you please help with the numbers? Maybe run in the reverse order?

@KlzXS I think this is happening as in the second run with shell spawning you are seeing the effect of caching. Try rebooting the system and running the tests in reverse order. shell spawning is generally visibly costly.

**Comment by KlzXS** [comment]
Now I get ~4s constantly even after reboot. Might have gotten lucky or unlucky with some system usage in the background. Either way as I've said spawning a shell is faster than not doing it. Just confirm that it's not only me getting such results.

**Comment by 0xACE** [comment]
> @0xACE can you please help with the numbers? Maybe run in the reverse order?

I have been too busy to keep up with `nnn` lately and will continue being busy until the end of this month.

Therefore I don't understand this problem nor can I give any helpful input until next sunday.

**Comment by jarun** [comment]
Next Sunday is fine. No hurry!

@KlzXS please continue with the detection logic. If the options are not available, the operation would be shell spawning so we don't have any other option I can think of.

**Comment by jarun** [comment]
@BKPepe please feel free to pitch in. I am on SSD on a very fast system and my results may not matter. Hence the request to others.

**Comment by 0xACE** [comment]
e-mail: sage

> Next Sunday is fine

Big oof, just realised, i meant 2019-12-01, as in december... and I can't predict what state I will be in at that time... anyhow the point is I'll be busy... I would've helped with the benchmark thing but producing an environment for it would require me to step out of my current responsibilities :(

**Comment by jarun** [comment]
No issues. We'll continue as planned. We have no xargs options. ;)

**Comment by KlzXS** [comment]
I ran the same test on Termux. Obviously my phone doesn't have an SSD so it took forever but still the same result.  Also I do have both `-I` and `-o` and `-a` available. 
```
time <list xargs -0 -o -I{} cp {} nnn/
real    38m21.319s
user    9m19.776s
sys     23m7.848s

time <list xargs -0 sh -c 'cp $0 $@ nnn/ </dev/tty'
real    5m25.261s
user    0m3.848s
sys     4m53.876s
```

I really would like to just go ahead and remove these flags completely. It would simplify the code and it doesn't worsen the performance, it improves it in this case.

**Comment by jarun** [comment]
Sure thing!

**Comment by jarun** [comment]
Please share another set of results on the same device - power off, run slow, then quick.

Whatever you run second will give better results due to caching. So best shot is cold boot, run quick then again cold boot, run slow. That is one reading per cold boot.

**Comment by KlzXS** [comment]
I'm not sure what cache you're referring to, I'm copying 400M worth of files containing random data. I am running the test, it'll be another half hour at least until it finishes. I put in a pull request with the changes anyways.

**Comment by KlzXS** [comment]
```
time <list xargs -0 -o -I{} cp {} nnn/
real    42m34.323s
user    10m52.144s
sys     26m3.512s

time <list xargs -0 sh -c 'cp $0 $@ nnn/ </dev/tty'
real    2m40.918s
user    0m2.468s
sys     2m37.340s
```

**Comment by jarun** [comment]
OK. I think we are good with: `<list xargs -0 sh -c 'cp $0 $@ nnn/ </dev/tty'` from the results.

Thank you for the data.

**Comment by jarun** [comment]
Closed by PR #388.
